core/src/main/java/org/bouncycastle/crypto/digests/Blake3Digest.java

```diff
@@ -0,0 +1,907 @@
+    private static final int KEYEDHASH     = 16;
+
+    /**
+     * DeriveContext Flag.
+     */
+    private static final int DERIVECONTEXT = 32;
+
+    /**
+     * DeriveKey Flag.
+     */
+    private static final int DERIVEKEY     = 64;
+
+    /**
+     * Chaining0 State Locations.
+     */
+    private static final int CHAINING0 = 0;
+
+    /**
+     * Chaining1 State Location.
+     */
+    private static final int CHAINING1 = 1;
+
+    /**
+     * Chaining2 State Location.
+     */
+    private static final int CHAINING2 = 2;
+
+    /**
+     * Chaining3 State Location.
+     */
+    private static final int CHAINING3 = 3;
+
+    /**
+     * Chaining4 State Location.
+     */
+    private static final int CHAINING4 = 4;
+
+    /**
+     * Chaining5 State Location.
+     */
+    private static final int CHAINING5 = 5;
+
+    /**
+     * Chaining6 State Location.
+     */
+    private static final int CHAINING6 = 6;
+
+    /**
+     * Chaining7 State Location.
+     */
+    private static final int CHAINING7 = 7;
+
+    /**
+     * IV0 State Locations.
+     */
+    private static final int IV0       = 8;
+
+    /**
+     * IV1 State Location.
+     */
+    private static final int IV1       = 9;
+
+    /**
+     * IV2 State Location.
+     */
+    private static final int IV2       = 10;
+
+    /**
+     * IV3 State Location.
+     */
+    private static final int IV3       = 11;
+
+    /**
+     * Count0 State Location.
+     */
+    private static final int COUNT0    = 12;
+
+    /**
+     * Count1 State Location.
+     */
+    private static final int COUNT1    = 13;
+
+    /**
+     * DataLen State Location.
+     */
+    private static final int DATALEN   = 14;
+
+    /**
+     * Flags State Location.
+     */
+    private static final int FLAGS     = 15;
+
+    /**
+     * Message word permutations.
+     */
+    private static final byte[] SIGMA = { 2, 6, 3, 10, 7, 0, 4, 13, 1, 11, 12, 5, 9, 14, 15, 8 };
+
+    /**
+     * Rotation constants.
+     */
+    private static final byte[] ROTATE = { 16, 12, 8, 7 };
+
+    /**
+     * Blake3 Initialization Vector.
+     */
+    private static final int[] IV = {
+            0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19
+    };
+
+    /**
+     * The byte input/output buffer.
+     */
+    private final byte[] theBuffer = new byte[BLOCKLEN];
+
+    /**
+     * The key.
+     */
+    private final int[] theK = new int[NUMWORDS];
+
+    /**
+     * The chaining value.
+     */
+    private final int[] theChaining = new int[NUMWORDS];
+
+    /**
+     * The state.
+     */
+    private final int[] theV = new int[NUMWORDS << 1];
+
+    /**
+     * The message Buffer.
+     */
+    private final int[] theM = new int[NUMWORDS << 1];
+
+    /**
+     * The indices.
+     */
+    private final byte[] theIndices = new byte[NUMWORDS << 1];
+
+    /**
+     * The chainingStack.
+     */
+    private final Deque<int[]> theStack = new ArrayDeque();
```

Comment:
Nit: this could specify an initial deque size of 54 since that's the max size the stack gets. Or a fixed size array to avoid additional garbage during hashing.